Feature/look back #patch

ShaneMPutnam opened the pull request:
This PR adds a date time selector component that is used to select the forecast publication time. The selected forecast publication time is then used to update the short and medium range forecast plot. Note, I also changed the plot time zone to local instead of UTC.

albrazeau commented:
@mxkpp The date picker was put in the header component instead of the plot component because when a date is selected, we would want the colors of the asset points on the map to update as well. I am currently looking into how to filter the results of pg_tileserv (i.e. based on a provided date), but I think we would need to update the hybrid materialized views, or turn them into functions that accept dates.

mxkpp commented:
> @mxkpp The date picker was put in the header component instead of the plot component because when a date is selected, we would want the colors of the asset points on the map to update as well. I am currently looking into how to filter the results of pg_tileserv (i.e. based on a provided date), but I think we would need to update the hybrid materialized views, or turn them into functions that accept dates.

@ShaneMPutnam @albrazeau Yes the MV logic would need to be turned into a function. There will be significant delay on that fetch though, as it would require calculating all forecasts for all points in the state, every time the date is changed by the user. E.g. currently refreshing Missouri's "hybrid nominal" MV takes about 8 seconds.

mxkpp commented:
The minimum date concept would break when we add new sites later to existing states.  The API would need to just omit sites that had no NWM data earlier than the given time. Could display a tailored message when the time is so early that no NWM data is available for any sites (instead of seeing an empty map wondering if it broke).

The main Postgres function is raising a custom error message to this effect in these two statements (not yet handled by API or client), but the second statement here would need to be changed since we would need to distinguish between missing NWM vs missing AHPS.

https://github.com/Dewberry/fcast-dashboard/blob/3de576abbb76ab6dfd836c17d31f1efcf52bae47/dba/create-schema.sql#L504

https://github.com/Dewberry/fcast-dashboard/blob/3de576abbb76ab6dfd836c17d31f1efcf52bae47/dba/create-schema.sql#L536

ShaneMPutnam commented:
> The minimum date concept would break when we add new sites later to existing states. The API would need to just omit sites that had no NWM data earlier than the given time. Could display a tailored message when the time is so early that no NWM data is available for any sites (instead of seeing an empty map wondering if it broke).
> 
> The main Postgres function is raising a custom error message to this effect in these two statements (not yet handled by API or client), but the second statement here would need to be changed since we would need to distinguish between missing NWM vs missing AHPS.
> 
> https://github.com/Dewberry/fcast-dashboard/blob/3de576abbb76ab6dfd836c17d31f1efcf52bae47/dba/create-schema.sql#L504
> 
> https://github.com/Dewberry/fcast-dashboard/blob/3de576abbb76ab6dfd836c17d31f1efcf52bae47/dba/create-schema.sql#L536

We could add a new endpoint to the API that is called the first time a trigger location is selected, which returns the minimum publication date for that point. This would prevent the map from being blank since there would always be at least one point with a forecast. However, yes, we would have to update the views/functions @albrazeau mentioned to omit those without forecasts.

mxkpp commented:
@ShaneMPutnam @albrazeau Following our meeting, changes to the map point layers have been tabled and moved into this new issue:
https://github.com/Dewberry/fcast-dashboard/issues/97

So please disregard my above comments in this PR as they pertain to the map layers, materialized views, points, etc.

ShaneMPutnam commented:
> @ShaneMPutnam @albrazeau Following our meeting, debate on changes to the map point layers have been tabled and moved into this new issue: #97
> 
> So please disregard my above comments in this PR as they pertain to the map layers, materialized views, points, etc.

@mxkpp, please take a look at my two most recent commits. I have addressed what we discussed earlier today, except for the date resetting to the current time when the chosen date is earlier than the minimum -- there does not seem to be an easy what to fix this.

mxkpp commented:
Looks great, thanks! I spoofed some earlier dates, including one so early there is no data, and all behaved as expected.
